cpp/src/lib/manifest_cache.cc

```diff
@@ -0,0 +1,49 @@
+/*
+Copyright (C) 2016 iNuron NV
+
+This file is part of Open vStorage Open Source Edition (OSE), as available from
+
+
+    http://www.openvstorage.org and
+    http://www.openvstorage.com.
+
+This file is free software; you can redistribute it and/or modify it
+under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
+as published by the Free Software Foundation, in version 3 as it comes
+in the <LICENSE.txt> file of the Open vStorage OSE distribution.
+
+Open vStorage is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY of any kind.
+*/
+
+#include "manifest_cache.h"
+
+namespace alba {
+namespace proxy_client {
+
+ManifestCache &ManifestCache::getInstance() {
+  static ManifestCache instance;
+  return instance;
+}
+
+void ManifestCache::add(strpair key, std::unique_ptr<Manifest> mfp) {
+  ALBA_LOG(DEBUG, "ManifestCache::add");
+
+  _cache_mutex.lock();
+
+  auto it = _cache.find(key);
+  if (it != _cache.end()) {
+    _cache.erase(it);
+  }
+  _cache.emplace(std::make_pair(key, std::move(mfp)));
+
+  _cache_mutex.unlock();
+}
```

Comment:
I think the lookup / remove / insert can be simplified to `_cache[key] = std::move(mfp);`.
But taking a step back (and without any test data yet), this global lock could become quite popular / cause a lot of contention. In that case (and btw, what about cache invalidation?) another layer of indirection could help: a map or sth. similar for the namespaces and then per namespace another map for the manifests. Each level with its own locking (R/W locks) / the namespace map could possibly even be done without locking and atomic ops instead in the read path.

NB: RAII (std::lock_guard) would be better for locking instead of explicit lock / unlock